### PR TITLE
Update broker config lib to be broker-aware

### PR DESCRIPTION
**Pull request by yolocs**
Fixes https://github.com/google/knative-gcp/issues/724

## Proposed Changes

- Group targets by broker
- Add functions to read/update broker
- Additional info for each broker (e.g. decouple queue)
- Renamed in-memory cache struct (`BasicTargets` -> `CachedTargets`)
- Make memory implementation thread safe
- Add external notify channel for file.Targets; this can be used to reconcile pubsub pull clients

**Release Note**

```release-note
- Group targets by broker
- Add functions to read/update broker
- Additional info for each broker (e.g. decouple queue)
- Renamed in-memory cache struct (`BasicTargets` -> `CachedTargets`)
- Make memory implementation thread safe
- Add external notify channel for file.Targets; this can be used to reconcile pubsub pull clients
```

### CODE REVIEW

**Comment by yolocs** [comment]
@grantr @ericlem

File: pkg/broker/config/cache_test.go [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+
+	var gotVal TargetsConfig
+	if err := proto.Unmarshal(gotBytes, &gotVal); err != nil {
+		t.Errorf("unexpected error from proto.Unmarshal: %v", err)
+	}
+	if !proto.Equal(&gotVal, val) {
+		t.Errorf("got unmarshaled targets=%+v, want=%+v", gotVal, val)
+	}
+
+	// Test EqualsBytes
```
Comment (ericlem) [inline pkg/broker/config/cache_test.go]: Maybe have a test where EqualsBytes returns false?

File: pkg/broker/config/cache_test.go [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+	targets := &CachedTargets{}
+	targets.Store(val)
+
+	gotStr := targets.String()
+	wantStr := val.String()
+	if gotStr != wantStr {
+		t.Errorf("BaseTargets.String() got=%s, want=%s", gotStr, wantStr)
+	}
+
+	// Test EqualsString
```
Comment (ericlem) [inline pkg/broker/config/cache_test.go]: Maybe a test where EqualsString returns false?

File: pkg/broker/config/memory/memory.go [hunk 3]
```diff
@@ -18,66 +18,115 @@ package memory
...
-		if _, ok := cfg.GetNamespaces()[target.GetNamespace()]; ok {
-			cfg.GetNamespaces()[target.GetNamespace()].GetNames()[target.GetName()] = &local
-		} else {
-			cfg.GetNamespaces()[target.GetNamespace()] = &config.NamespacedTargets{
-				Names: map[string]*config.Target{
-					target.GetName(): &local,
-				},
-			}
+
+	if newVal != nil && newVal.Brokers != nil {
```
Comment (ericlem) [inline pkg/broker/config/memory/memory.go]: Is there any way newVal can be nil at this point?  it looks like you either made a copy of a non-nil value or explicitly allocated a new structure immediately above.

**Comment by ericlem** [comment]
/hold
/lgtm
I had a couple nits.  Up to you if you want to address.  Also want to give @grantr a chance to look since he'll be directly using it soon.

**Comment by grantr** [comment]
/unhold

No need to hold this for my review; we can make changes as we go. I'll be working on the configmap-writing part of the controller this week so I want this merged. :)

**Pull request merged**